modules/renter/proto/merkleroots.go

```diff
@@ -0,0 +1,369 @@
+	// by 1 root.
+	if err := mr.moveLastCachedSubTreeToUncached(); err != nil {
+		return err
+	}
+	return nil
+}
+
+// insert inserts a root by replacing a root at an existing index.
+func (mr *merkleRoots) insert(index int, root crypto.Hash) error {
+	if index > mr.numMerkleRoots {
+		return errors.New("can't insert at a index greater than the number of roots")
+	}
+	if index == mr.numMerkleRoots {
+		return mr.push(root)
+	}
+	// Replaced the root on disk.
+	_, err := mr.rootsFile.WriteAt(root[:], fileOffsetFromRootIndex(index))
+	if err != nil {
+		return errors.AddContext(err, "failed to insert root on disk")
+	}
+
+	// Find out if the root is in mr.cachedSubTree or mr.uncachedRoots.
+	i, cached := mr.isIndexCached(index)
+	// If the root was not cached we can simply replace it in mr.uncachedRoots.
+	if !cached {
+		mr.uncachedRoots[i] = root
+		return nil
+	}
+	// If the root was cached we need to rebuild the cache.
+	if err := mr.rebuildCachedTree(i); err != nil {
+		return errors.AddContext(err, "failed to rebuild cache for inserted root")
+	}
+	return nil
+}
+
+// isIndexCached determines if the root at index i is already cached in
+// mr.cachedSubTree or if it is still in mr.uncachedRoots. It will return true
+// or false and the index of the root in the corresponding data structure.
+func (mr *merkleRoots) isIndexCached(i int) (int, bool) {
+	if i/merkleRootsPerCache == len(mr.cachedSubTrees) {
+		// Root is not cached. Return the false and the position in
+		// mr.uncachedRoots
+		return i - len(mr.cachedSubTrees)*merkleRootsPerCache, false
+	}
+	return i / merkleRootsPerCache, true
+}
+
+// lenFromFile returns the number of merkle roots by computing it from the
+// filesize.
+func (mr *merkleRoots) lenFromFile() (int, error) {
+	size, err := mr.rootsFile.Size()
+	if err != nil {
+		return 0, err
+	}
+
+	// Sanity check contract file length.
+	if size%crypto.HashSize != 0 {
+		return 0, errors.New("contract file has unexpected length and might be corrupted")
+	}
+	return int(size / crypto.HashSize), nil
+}
+
+// len returns the number of merkle roots. It should always return the same
+// number as lenFromFile.
+func (mr *merkleRoots) len() int {
+	return mr.numMerkleRoots
+}
+
+// moveLastCachedSubTreeToUncached deletes the last cached subTree and appends
+// its elements to the uncached roots.
+func (mr *merkleRoots) moveLastCachedSubTreeToUncached() error {
+	mr.cachedSubTrees = mr.cachedSubTrees[:len(mr.cachedSubTrees)-1]
+	rootIndex := len(mr.cachedSubTrees) * merkleRootsPerCache
+	roots, err := mr.merkleRootsFromIndexFromDisk(rootIndex, mr.numMerkleRoots)
+	if err != nil {
+		return errors.AddContext(err, "failed to read cached tree's roots")
+	}
+	mr.uncachedRoots = append(mr.uncachedRoots, roots...)
+	return nil
+}
+
+// push appends a merkle root to the end of the contract. If the number of
+// uncached merkle roots grows too big we cache them in a new subTree.
+func (mr *merkleRoots) push(root crypto.Hash) error {
```

Comment:
`push` has the same problem as delete - it's not idempotent. WAL may end up calling an update multiple times if there are crashes at key places during disk management, so we need to restructure all of these functions to be idempotent.

This means we also need to write idempotency tests to verify that all of the operations are being updated idempotently. In the host we do this by adding a disrupt during WAL application that might write an update multiple times.

True idempotency though would be to apply some of the transactions in a WAL, perhaps even get partway through one of the transactions, and then apply them again a few times, each time stopping at a different update (or even finishing all the way), until finally on the last one you just go all the way.

If the WAL is built correctly, the result should be identical to just applying the transaction once.